### ISSUE TITLE
Predict results by java and python api are not exactly same?

### ISSUE THREAD

**Issue by Luminosite** [tensorflow/tensorflow#43560]
Hi, I'm trying to run tensorflow model on spark, and testing the java api. I built a simple model by tensorflow and tried to run it with both java and python API. I found many results are not exactly same, but with different bit precision, and java results' precision bits are usually less then python's, for example:

```
java results:
       [9.9787010e-01, 1.7358017e-05, 2.1125420e-03],
       [9.9581650e-01, 8.3105200e-04, 3.3524006e-03],
       [9.9857880e-01, 3.4846460e-05, 1.3863753e-03],
       [9.9834824e-01, 3.1828644e-05, 1.6199955e-03],
       [9.9810120e-01, 2.9924562e-05, 1.8689097e-03],
       [9.8520744e-01, 7.6777805e-03, 7.1147494e-03],

python results:
       [9.9787009e-01, 1.7358017e-05, 2.1125420e-03],
       [9.9581653e-01, 8.3105202e-04, 3.3524006e-03],
       [9.9857879e-01, 3.4846460e-05, 1.3863753e-03],
       [9.9834824e-01, 3.1828644e-05, 1.6199955e-03],
       [9.9810117e-01, 2.9924562e-05, 1.8689097e-03],
       [9.8520744e-01, 7.6777805e-03, 7.1147499e-03],
```

The example I used is iris classification, results are iris's three types, so I used one-hot code.
The most wired thing is that java results' precision bits are less then python's obviously. So, is there any config or any way to make sure java result get same precision with python?

The codes I used to generate model:
```
def create_model():
    inputs = tf.keras.Input(shape=(4,), name='k_input')
    x = tf.keras.layers.Dense(10, activation=tf.nn.relu)(inputs)
    outputs = tf.keras.layers.Dense(3, activation=tf.nn.softmax, name='out')(x)
    model = tf.keras.Model(inputs=inputs, outputs=outputs)

    model.compile(optimizer='adam',
                loss=tf.losses.SparseCategoricalCrossentropy(from_logits=True),
                metrics=[tf.metrics.SparseCategoricalAccuracy()])

    return model


model = create_model()
model.fit(vec_train, res_train, epochs=10)
model.save("saved_model/iris")
```
The codes I used to run model in python:
```
loaded = keras.models.load_model("saved_model/iris")
res = loaded.predict(vec_sample)
```
The codes I used in java:
```
    val model = SavedModelBundle.load(s"$resourceDir/iris", "serve")
    val session = model.session()

    val ret = session
      .runner()
      .feed("serving_default_k_input", Tensor.create(inputData))
      .fetch("StatefulPartitionedCall")
      .run()
    val size = inputData.length * 3
    val r = FloatBuffer.allocate(size.toInt)
    val result = ret.get(0).writeTo(r)
```

tensorflow version: 2.1
python version: 3.7
java version: 1.8
java api version: 1.15.0

**Comment by bhack** [comment]
Can you try to print the inputs in python and java?

**Comment by jvishnuvardhan** [comment]
@Luminosite I have couple of questions.
1) Is there any model conversion involved here? 
2) I checked the above results. 2nd and third columns are exactly same. Only change is in first column of results and the change is in 6th or 7 decimal. Do you want 6th decimal accuracy for your use-case? Is there any rounding error involved?
3) is there any difference in the predictions? Thanks!

I guess the discrepancy is very very low or insignificant. But, may be your use-case needs that accuracy. Thanks!

**Comment by Luminosite** [comment]
> Can you try to print the inputs in python and java?

```
input: [7.174808,2.6553774,5.12266,0.41229987], output: [0.9990282,3.3963486E-5,9.378085E-4]
input: [7.191363,2.555118,6.765061,2.0554824], output: [0.99972814,4.5734428E-6,2.6736E-4]
input: [6.6439853,3.5317183,5.3618083,1.0571597], output: [0.99917054,2.1115382E-5,8.0844515E-4]
input: [6.6596174,2.955047,5.980057,1.1314213], output: [0.999271,1.4503375E-5,7.145047E-4]
input: [5.7817345,4.1580234,3.0035162,0.73943543], output: [0.99696285,4.2656367E-4,0.0026106567]
input: [5.461476,4.3531785,5.786151,1.8375901], output: [0.9990048,1.573877E-5,9.795206E-4]
input: [7.2582817,3.4890115,4.6998153,2.4146466], output: [0.999686,3.089602E-5,2.8309433E-4]
```
and I used same data directly in python to run the model like:
```
input_sample = [
    [7.174808,2.6553774,5.12266,0.41229987],
    [7.191363,2.555118,6.765061,2.0554824],
    [6.6439853,3.5317183,5.3618083,1.0571597],
    ... ...
]
loaded.predict(input_sample)
```
finally I got result with more precision bits:
```
array([[9.99028206e-01, 3.39634862e-05, 9.37808480e-04],
       [9.99728143e-01, 4.57344277e-06, 2.67360010e-04],
       [9.99170542e-01, 2.11153820e-05, 8.08445155e-04],
       [9.99270976e-01, 1.45033746e-05, 7.14504684e-04],
       [9.96962845e-01, 4.26563667e-04, 2.61065667e-03],
       [9.99004781e-01, 1.57387694e-05, 9.79520613e-04],
       [9.99686003e-01, 3.08960189e-05, 2.83094327e-04],
       ... ...
])
```

**Comment by Luminosite** [comment]
> @Luminosite I have couple of questions.
> 
> 1. Is there any model conversion involved here?
> 2. I checked the above results. 2nd and third columns are exactly same. Only change is in first column of results and the change is in 6th or 7 decimal. Do you want 6th decimal accuracy for your use-case? Is there any rounding error involved?
> 3. is there any difference in the predictions? Thanks!
> 4. can you please use recent TF versions and let us know how those perform.
> 
> I guess the discrepancy is very very low or insignificant. But, may be your use-case needs that accuracy. Thanks!

Thanks for the reply.
Here're my answers:
1, no, I only used a NN like the codes above.
2, we want to provide model prediction service by spark, so we need to make sure the results are as precise as python results.  And I didn't involve any rounding process, so I guess it should not be the rounding problem
3, results are almost same, all differeces are at 10^-7 level, so I think it's caused by different number of precision bits
4, I've used latest tensorflow dependency in maven, which is 1.15.0. I couldn't find any dependencies later than 1.15.0 in https://repo.maven.apache.org/maven2/org/tensorflow/tensorflow/

And also, I noticed something may related to the issue. In java codes, I have to convert input data as float, and result data is also of type float. In java, float type is only 34 bits precision which support around 6~7 bits precision in decimalism. But in python the default type of float number is 'float' which is not same in java but has more precision bits.
I wonder if there're some more precision API in java to provide more precise result?
